### PR TITLE
SERENE-748: Read the dataset and parse into EPG objects

### PR DESCRIPTION
**Pull request by gaj67**

### CODE REVIEW

**Comment by gaj67** [comment]
I've just noticed that this AC has not been done:
* Update the spring EpgManagementService for reading EPGs

Do we want it as part of this task, or should it go into SERENE-749 or later? This is an integration issue rather than a functionality issue.

File: search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestor.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+    private void ingestGraphHeads(URI graphData) {
+        if (graphData != null && "file".equals(graphData.getScheme())) {
+            new EpgElementFileIngestor<EpgHead>(graphSet, new EpgHeadExtractor()).ingest(getFilePath(graphData));
+        } else {
+            throw new EpgDataException("Unhandled scheme for URI: " + graphData);
+        }
+    }
+
+    private Path getFilePath(URI uriPath) {
```
Comment (tmoschou) [inline search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestor.java]: I'ld just inline this method with `java.nio.file.Paths.get(uriPath)`

File: search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgElementsMap.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+public class EpgElementsMap<T extends EpgElement> implements EpgElementContainer<T> {
+
+    private final Map<String, T> map = new HashMap<>();
+
+    @Override
+    public void addElement(T elem) throws EpgElementExistsException {
+        T oldElem = map.put(elem.getId(), elem);
+        if (oldElem != null) {
+            throw new EpgElementExistsException(
```
Comment (tmoschou) [inline search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgElementsMap.java]: hmm, if we are throwing this exception, we should probably not modify the internal state, which has already happened. Should check key exists first before putting

File: search/src/main/java/au/com/d2dcrc/ia/search/ingestor/extractor/EpgEdgeExtractor.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * Allows the repeated extraction of EPG edges from JSON strings in
+ * {@link EpgEdgeModel} format.
+ */
+public class EpgEdgeExtractor extends EpgElementExtractor<EpgEdge> {
+
+    @Override
+    public EpgEdge extract(String json) throws EpgSyntaxException {
+        try {
+            EpgEdgeModel model = mapper.readValue(json, EpgEdgeModel.class);
+            model.validate();
```
Comment (tmoschou) [inline search/src/main/java/au/com/d2dcrc/ia/search/ingestor/extractor/EpgEdgeExtractor.java]: We should use JSR 303/349/380 bean validation for this, instead of writing custom validation methods since our models are already annotated. As simple as injecting a `javax.validation.Validator validator` somewhere and running
```java
Set<ConstraintViolation<String>> violations = validator.validate(model);
if (!violations.isEmpty()) {
    throw new ConstraintViolationException(violations);
}
```

File: search/src/test/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestorTest.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+public class EpgDataIngestorTest {
+
+    /**
+     * Tests against the IMBD flat data set.
+     * 
+     * @throws IOException if something bad happens.
+     * @throws URISyntaxException if the resource path fails.
+     */
+    @Test
+    public void testIMDBIngestion() throws IOException, URISyntaxException {
```
Comment (tmoschou) [inline search/src/test/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestorTest.java]: `testImdbIngestion()`. Probably should add a backlog task to have checkstyle run over the test classes, too.

File: search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgElementFileIngestor.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            ingestElementFromDirectory(path);
+        } else if (Files.isRegularFile(path)) {
+            ingestElementFromFile(path);
+        } else {
+            throw new EpgDataException("Invalid ingestion path: " + path);
+        }
+    }
+
+    private void ingestElementFromDirectory(Path path) {
+        try (Stream<Path> files = Files.walk(path).filter(p -> Files.isRegularFile(p))) {
```
Comment (tmoschou) [inline search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgElementFileIngestor.java]: this makes me so happy :)

Comment (tmoschou) [reply search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgElementFileIngestor.java]: I.e. Using streams and using the nio.

File: search/src/main/java/au/com/d2dcrc/ia/search/ingestor/extractor/EpgEdgeExtractor.java (hunk 3 above)
Comment (tmoschou) [reply search/src/main/java/au/com/d2dcrc/ia/search/ingestor/extractor/EpgEdgeExtractor.java]: Since these aren't managed by strings, you'll prob need to inject the validator to some class in the upper layer, and pass in yourself as a constructor arg. Else you could make these spring components if they are singleton (or stateless)

Comment (tmoschou) [reply search/src/main/java/au/com/d2dcrc/ia/search/ingestor/extractor/EpgEdgeExtractor.java]: It will also mean you will be able to reuse the preconfigured, object mapper rather than construct a new one

File: search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestor.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import au.com.d2dcrc.ia.search.ingestor.extractor.EpgVertexExtractor;
+import au.com.d2dcrc.ia.search.management.EpgReferenceModel;
+import java.io.File;
+import java.net.URI;
+import java.nio.file.Path;
+
+/**
+ * Ingests EPG graphs from raw data.
+ */
+public class EpgDataIngestor {
```
Comment (bowernj) [inline search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestor.java]: Should this class be a Bean and managed by Spring?

File: search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestor.java (hunk 1 above)
Comment (gaj67) [reply search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestor.java]: How do you want the possible exceptions of Paths.get() handled?

File: search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgElementsMap.java (hunk 2 above)
Comment (gaj67) [reply search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgElementsMap.java]: Done.

File: search/src/test/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestorTest.java (hunk 4 above)
Comment (gaj67) [reply search/src/test/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestorTest.java]: Thought it already did - no wonder some long lines were allowed!

Comment (tmoschou) [reply search/src/test/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestorTest.java]: Done: https://jira.csiro.au/browse/SERENE-840

File: search/src/test/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestorTest.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+/**
+ * Tests the implementation of {@link EpgDataIngestor}.
+ */
+public class EpgDataIngestorTest {
+
+    private final ObjectMapper mapper = new ObjectMapper();
+    private final ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory();
+    private final Validator validator = validatorFactory.usingContext().getValidator();
+    private final EpgDataIngestor ingestor = new EpgDataIngestor(mapper, validator);
```
Comment (bowernj) [inline search/src/test/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestorTest.java]: Should now be able to use the managed EpgDataIngestor bean, rather than new.

File: search/src/test/java/au/com/d2dcrc/ia/search/ingestor/extractor/EpgEdgeExtractorTest.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+import javax.validation.Validator;
+import javax.validation.ValidatorFactory;
+import org.junit.Test;
+
+/**
+ * Tests the implementation of {@link EpgEdgeExtractor}.
+ */
+public class EpgEdgeExtractorTest {
+
+    private final ObjectMapper mapper = new ObjectMapper();
```
Comment (bowernj) [inline search/src/test/java/au/com/d2dcrc/ia/search/ingestor/extractor/EpgEdgeExtractorTest.java]: Use the injected ObjectMapper for all Extractor and Model tests?

Comment (gaj67) [reply search/src/test/java/au/com/d2dcrc/ia/search/ingestor/extractor/EpgEdgeExtractorTest.java]: The question I still have for unit tests is: do we want (or require) the application to run for the test? In these cases I don't think we do - the ObjectMapper doesn't need to handle any special values (e.g. date conversion), so a vanilla version will do.

File: search/src/test/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestorTest.java (hunk 7 above)
Comment (gaj67) [reply search/src/test/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestorTest.java]: See my comment below. Also, the bean injection is tested via createEpg, so it doesn't require a duplicate test here, I feel.

File: search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestor.java (hunk 6 above)
Comment (gaj67) [reply search/src/main/java/au/com/d2dcrc/ia/search/ingestor/EpgDataIngestor.java]: Done.
